worker/rhpv3.go

```diff
@@ -551,111 +507,149 @@ func readSectorCost(pt *rhpv3.HostPriceTable) (types.Currency, error) {
+	if ongoing && time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway)) {
+		return
+	}
+
+	// price table is being updated, wait for the update
+	if ongoing {
 		select {
 		case <-ctx.Done():
-			return rhpv3.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
-		case <-ongoing.done:
-		}
-		if ongoing.err != nil {
-			return rhpv3.HostPriceTable{}, ongoing.err
-		} else {
-			return *ongoing.pt, nil
+			return hostdb.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
+		case <-update.done:
 		}
+		return update.hpt, update.err
 	}
 
-	// Update price table.
-	var hpt rhpv3.HostPriceTable
-	err := withTransportV3(ctx, siamuxAddr, hk, func(t *rhpv3.Transport) (err error) {
-		hpt, err = RPCPriceTable(t, payFn)
-		return err
-	})
+	// this thread is updating the price table
+	defer func() {
+		update.hpt = hpt
+		update.err = err
+		close(update.done)
+
+		p.mu.Lock()
+		p.hpt = hpt
+		p.update = nil
+		p.mu.Unlock()
+	}()
 
-	pt.mu.Lock()
-	defer pt.mu.Unlock()
+	// fetch the host info from the bus, it might have a new price table already
+	// in which case we can return without having to update it
+	host, err := b.Host(ctx, hk)
+	if err == nil && host.PriceTable != nil && time.Now().Before(host.PriceTable.Expiry.Add(priceTableValidityLeeway)) {
+		hpt = *host.PriceTable
+		return
+	}
+	hostIP := host.NetAddress
+	siamuxAddr := host.Settings.SiamuxAddr()
 
-	// On success we update the pt.
-	if err == nil {
-		ongoing.pt = &hpt
-		pt.pt = &hpt
-		pt.expiry = time.Now().Add(hpt.Validity)
+	payByFC := func() (pt rhpv3.HostPriceTable, err error) {
+		// if revision was given, use it to pay for the scan
+		if revision != nil {
+			_, _, pt, err = w.scanHost(ctx, hk, hostIP, siamuxAddr, w.preparePriceTableContractPayment(hk, revision))
+			return
+		}
+
+		// otherwise we fetch the active contract
+		contract, err := b.ActiveContract(ctx, hk)
+		if err != nil {
+			return rhpv3.HostPriceTable{}, err
+		}
+
+		// and scan the host
+		_ = w.withRevision(ctx, contract.ID, hk, hostIP, lockingPriorityPriceTable, lockingDurationPriceTable, func(revision types.FileContractRevision) error {
+			_, _, pt, err = w.scanHost(ctx, hk, hostIP, siamuxAddr, w.preparePriceTableContractPayment(hk, &revision))
+			return err
+		})
+		return
 	}
-	// Signal that the update is over.
-	ongoing.err = err
-	close(ongoing.done)
-	pt.ongoingUpdate = nil
-	return hpt, err
-}
 
-// priceTable returns a priceTable from priceTables for the given host or
-// creates a new one.
-func (pts *priceTables) priceTable(hk types.PublicKey) *priceTable {
-	pts.mu.Lock()
-	defer pts.mu.Unlock()
+	// scan host and pay by EA if possible
+	var pt rhpv3.HostPriceTable
+	if revision != nil {
+		pt, err = payByFC() // pay by FC if a revision is given
+	} else if cs, errr := b.ConsensusState(ctx); errr == nil && cs.Synced {
+		_, _, pt, err = w.scanHost(ctx, hk, hostIP, siamuxAddr, w.preparePriceTableAccountPayment(hk, cs.BlockHeight))
```

Comment:
Merging fetching the settings and fetching the pricetable within the worker feels a bit strange tbh.
Because we are talking about 2 different protocols here and price table code should never need to use rhpv2. 
Especially after merging #224.